### ISSUE TITLE
etcd proxy (2.0.9) leaving thousands of tcp connections in CLOSE_WAIT state

### ISSUE THREAD

**Issue by qix** [coreos/etcd#2743]
I've recently set up an etcd cluster to serve kubernetes (on the same hosts as my current consul cluster). The cluster has five members, and I have a single etcd box running with `--proxy on` which is receiving all the queries from the master server.

I'm seeing an issue where the the number of open files by etcd slowly climbs (leaking?) until it hits the file limit and then stops serving requests with "2015/04/22 03:59:10 http: Accept error: accept tcp [::]:2379: too many open files; retrying in 1s"

I'm not sure how to reproduce this. Here is some output that might help?

```
# cat /etc/supervisor.d/etcd.conf | grep command
command=sh -c "ulimit -n 16384; /usr/local/bin/etcd --advertise-client-urls http://10.240.178.193:2379 --listen-client-urls http://0.0.0.0:2379 --proxy on --initial-cluster <...> --data-dir /data/etcd --name k8s-master1"

# lsof | grep etcd.*CLOSE_WAIT | head
etcd      30980             root    7u     IPv6           10943418      0t0        TCP localhost:2379->localhost:42973 (CLOSE_WAIT)
etcd      30980             root    8u     IPv6           10859166      0t0        TCP localhost:2379->localhost:49911 (CLOSE_WAIT)
etcd      30980             root   11u     IPv6           10901434      0t0        TCP localhost:2379->localhost:33732 (CLOSE_WAIT)
etcd      30980             root   13u     IPv6           10645832      0t0        TCP localhost:2379->localhost:40868 (CLOSE_WAIT)
etcd      30980             root   14u     IPv6           10848580      0t0        TCP localhost:2379->localhost:46664 (CLOSE_WAIT)

# lsof | grep etcd.*CLOSE_WAIT | wc -l
2835

# etcdctl --version
etcdctl version 2.0.9

# etcdctl member list
5db7392a64a763b: name=consul2 peerURLs=http://consul2:2380 clientURLs=http://10.240.73.76:2379
fa87a16f1f0e26d: name=consul1 peerURLs=http://consul1:2380 clientURLs=http://10.240.86.161:2379
4db1c9b7b2aa4bf2: name=consul5 peerURLs=http://consul5:2380 clientURLs=http://10.240.89.41:2379
ae3317f6e55e4b5f: name=consul3 peerURLs=http://consul3:2380 clientURLs=http://10.240.30.116:2379
bb30e8aae19e352d: name=consul4 peerURLs=http://consul4:2380 clientURLs=http://10.240.247.33:2379

# etcdctl cluster-health
cluster is healthy
member 4db1c9b7b2aa4bf2 is healthy
member 5db7392a64a763b is healthy
member ae3317f6e55e4b5f is healthy
member bb30e8aae19e352d is healthy
member fa87a16f1f0e26d is healthy
```

**Comment by xiang90** [comment]
@qix Do you have a lot of watchers? What does your proxy workload look like?

**Comment by qix** [comment]
I'm not sure to be honest. The only thing hitting the etcd daemon is a kubernetes master process (scheduler, api, etc.)

Is the 3000 open files a concern? If not we can probably close this issue. After bumping up the file limit it seems to be running alright; with a maximum of 4000 open files.

It does seem a little strange that thousands of files are in the `CLOSE_WAIT` state on etcd. If that is expected though with the bumped file limit everything seems to be running swimmingly on my side.

**Comment by yichengq** [comment]
Current underlying transport in proxy recycles connections quickly regardless of new requests are still coming, which results CLOSE_WAIT-state connections.

I built a 3-member cluster on GCE, start a proxy on the 4th machine, and use 200 clients to use proxy do write requests. The peak of fd usage is around 600. I think the peak value depends on both the number of concurrent clients and how long system recycles closed connections.

The behavior should be improved to use fewer open files.

I don't know whether the behavior makes sense because i don't know the number of concurrent clients. Do you have any rough estimation of that?

Moreover, metrics on proxy ops can be really helpful to debug this case. and we are doing it here: https://github.com/coreos/etcd/pull/2727

**Comment by xiang90** [comment]
fixed by #2771
